src/module.c

```diff
@@ -6461,6 +6461,24 @@ const char *RM_CallReplyProto(RedisModuleCallReply *reply, size_t *len) {
     return callReplyGetProto(reply, len);
 }
 
+int RM_EndExecutionUnit() {
+    /* should only execute at top level of nesting, as otherwise would cause changes to impact DB before top level
+     * expects it.
```

Comment:
Can you explain why the nesting level is relevant here? What risk can happened if nesting level is more then one that can not happened if nesting level is 1?

Reply:
imagine we have 2 modules that use this API.  Module X calls Module Y via an RM_Call (and other things).  Module X expects no side effects (i.e. its a transaction until it ends/drains the execution unit).   If Module Y is able to drain and cause side effects, module X no longer has that knowledge of what is going on.

as a simple example.

some client elsewhere brpop x

module X

RM_Call(lpush x 1)
RM_Call(module-y command) // nothing to do with 'x'
RM_Call(lrange x 0 -1)
RM_EndExecutionUnit()

module X might expect that since module Y's command doesn't have anything to do with key 'x', that nothing external will occur to change 'x'.  but if we can nest and this occurs, we now have a problem.

With that said, one could argue that now with the GIL modification, that same thing could happen, so that expectation can no longer be justified and we can remove that check from here.

i.e. a redis module that runs on the main thread (so doesn't use the GIL) which calls a function on a module that uses threads and hence does use the GIL.  taking/releasing the GIL will unblock pending clients.  

on the flip side, module Y will actually assert on trying to take the GIL if called in this way, as the assert on server.execution_nesting == 0 will be false, so we can't call a GIL based module from a module that runs on the main thread.